File: references/betterlectio/android/app/src/main/java/dk/betterlectio/android/feature/supabase/SupabaseFeedbackService.kt
package dk.betterlectio.android.feature.supabase

import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.rpc
import io.github.jan.supabase.storage.storage
import io.ktor.http.ContentType
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import timber.log.Timber
import java.util.UUID
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Submit private feedback via `submit_feedback` + Storage attachment upload.
 */
@Singleton
class SupabaseFeedbackService @Inject constructor(
    private val manager: SupabaseManager,
) {
    data class SubmitRequest(
        val studentId: String,
        val schoolId: Int,
        val category: String,
        val message: String,
        val context: Map<String, Any?>,
        val screenshotJpeg: ByteArray? = null,
        val screenshotWidth: Int? = null,
        val screenshotHeight: Int? = null,
    )

    data class SubmitResult(
        val feedbackId: String,
        val attachmentId: String? = null,
    )

    suspend fun submit(request: SubmitRequest): SubmitResult {
        val client = manager.client
            ?: error("Supabase not configured")
        val session = manager.awaitSessionReady()
        if (session is SupabaseSessionState.Unavailable) {
            throw SupabaseUnavailableException(session.reason)
        }

        val contextJson = buildJsonObject {
            for ((key, value) in request.context) {
                when (value) {
                    null -> Unit
                    is Boolean -> put(key, JsonPrimitive(value))
                    is Number -> put(key, JsonPrimitive(value))
                    else -> {
                        val text = value.toString()
                        if (text.isNotBlank()) put(key, JsonPrimitive(text))
                    }
                }
            }
        }

        val feedbackId = client.postgrest.rpc(
            function = "submit_feedback",
            parameters = SubmitFeedbackParams(
                pStudentId = request.studentId,
                pSchoolId = request.schoolId,
                pCategory = request.category,
                pMessage = request.message,
                pPlatform = "android",
                pContext = contextJson,
            ),
        ).decodeAs<String>()

        Timber.i("feedback submitted id=%s", feedbackId)

        var attachmentId: String? = null
        val jpeg = request.screenshotJpeg
        if (jpeg != null && jpeg.isNotEmpty()) {
            val path =
                "${request.schoolId}/${request.studentId}/$feedbackId/${UUID.randomUUID()}.jpg"
            try {
                client.storage.from(BUCKET).upload(path, jpeg) {
                    upsert = false
                    contentType = ContentType.Image.JPEG
                }
                attachmentId = client.postgrest.rpc(
                    function = "register_feedback_attachment",
                    parameters = RegisterAttachmentParams(
                        pFeedbackId = feedbackId,
                        pKind = "screenshot",
                        pStoragePath = path,
                        pMimeType = "image/jpeg",
                        pByteSize = jpeg.size,
                        pWidth = request.screenshotWidth,
                        pHeight = request.screenshotHeight,
                    ),
                ).decodeAs<String>()
                Timber.i("feedback attachment registered id=%s path=%s", attachmentId, path)
            } catch (t: Throwable) {
                // Feedback row already exists — don't fail the whole submit on attachment.
                Timber.e(t, "feedback attachment upload/register failed path=%s", path)
            }
        }

        return SubmitResult(feedbackId = feedbackId, attachmentId = attachmentId)
    }

    @Serializable
    private data class SubmitFeedbackParams(
        @SerialName("p_student_id") val pStudentId: String,
        @SerialName("p_school_id") val pSchoolId: Int,
        @SerialName("p_category") val pCategory: String,
        @SerialName("p_message") val pMessage: String,
        @SerialName("p_platform") val pPlatform: String,
        @SerialName("p_context") val pContext: JsonObject,
    )

    @Serializable
    private data class RegisterAttachmentParams(
        @SerialName("p_feedback_id") val pFeedbackId: String,
        @SerialName("p_kind") val pKind: String,
        @SerialName("p_storage_path") val pStoragePath: String,
        @SerialName("p_mime_type") val pMimeType: String? = null,
        @SerialName("p_byte_size") val pByteSize: Int? = null,
        @SerialName("p_width") val pWidth: Int? = null,
        @SerialName("p_height") val pHeight: Int? = null,
    )

    companion object {
        const val BUCKET = "feedback-attachments"
    }
}

File: references/betterlectio/ios/BetterLectio/SupabaseFeedbackService.swift
import Foundation
import Supabase

@MainActor
protocol FeedbackSubmitting {
    func submit(
        student: Student,
        category: FeedbackCategory,
        message: String,
        context: FeedbackContext,
        screenshot: EncodedFeedbackScreenshot?
    ) async throws -> FeedbackSubmitResult
}

@MainActor
final class SupabaseFeedbackService: FeedbackSubmitting {
    static let shared = SupabaseFeedbackService()

    private let manager: SupabaseManager

    init(manager: SupabaseManager = .shared) {
        self.manager = manager
    }

    func submit(
        student: Student,
        category: FeedbackCategory,
        message: String,
        context: FeedbackContext,
        screenshot: EncodedFeedbackScreenshot?
    ) async throws -> FeedbackSubmitResult {
        guard !student.isDemo else { throw FeedbackSubmissionError.demoMode }
        guard let client = manager.client else { throw FeedbackSubmissionError.notConfigured }
        guard !message.isEmpty else { throw FeedbackSubmissionError.emptyMessage }
        guard message.count <= FeedbackLimits.maximumMessageCharacters else {
            throw FeedbackSubmissionError.messageTooLong
        }

        if client.auth.currentSession == nil {
            guard
                let credentials = KeychainManager.shared.loadCredentials(for: student.studentId),
                await SupabaseAuthService.shared.ensureSession(
                    credentials: credentials,
                    studentId: student.studentId,
                    gymId: student.gymId
                )
            else {
                throw FeedbackSubmissionError.notAuthenticated
            }
        }

        let feedbackID: String = try await client.rpc(
            "submit_feedback",
            params: SubmitFeedbackParameters(
                p_student_id: student.studentId,
                p_school_id: student.gymId,
                p_category: category.rawValue,
                p_message: message,
                p_platform: "ios",
                p_context: context
            )
        )
        .execute()
        .value

        let attachmentStatus: FeedbackAttachmentStatus
        if let screenshot {
            attachmentStatus = await uploadScreenshot(
                screenshot,
                feedbackID: feedbackID,
                student: student,
                client: client
            )
        } else {
            attachmentStatus = .notRequested
        }

        return FeedbackSubmitResult(
            feedbackID: feedbackID,
            attachmentStatus: attachmentStatus,
            isDemo: false
        )
    }

    private func uploadScreenshot(
        _ screenshot: EncodedFeedbackScreenshot,
        feedbackID: String,
        student: Student,
        client: SupabaseClient
    ) async -> FeedbackAttachmentStatus {
        let path = "\(student.gymId)/\(student.studentId)/\(feedbackID)/\(UUID().uuidString.lowercased()).jpg"
        var didUpload = false
        do {
            try await client.storage
                .from("feedback-attachments")
                .upload(
                    path,
                    data: screenshot.data,
                    options: FileOptions(contentType: "image/jpeg", upsert: false)
                )
            didUpload = true

            let _: String = try await client.rpc(
                "register_feedback_attachment",
                params: RegisterFeedbackAttachmentParameters(
                    p_feedback_id: feedbackID,
                    p_kind: "screenshot",
                    p_storage_path: path,
                    p_mime_type: "image/jpeg",
                    p_byte_size: screenshot.data.count,
                    p_width: screenshot.width,
                    p_height: screenshot.height
                )
            )
            .execute()
            .value
            return .uploaded
        } catch {
            FeedbackLogBuffer.shared.record("Feedback screenshot upload failed: \(type(of: error))", level: "W")
            if didUpload {
                do {
                    try await client.storage.from("feedback-attachments").remove(paths: [path])
                } catch {
                    FeedbackLogBuffer.shared.record("Feedback screenshot cleanup failed: \(type(of: error))", level: "W")
                }
            }
            // The text feedback already exists. Report partial success so retrying
            // cannot accidentally create a duplicate feedback row.
            return .failed
        }
    }
}

struct EncodedFeedbackScreenshot: Sendable {
    let data: Data
    let width: Int
    let height: Int
}

struct FeedbackContext: Encodable, Sendable {
    let app_version: String
    let app_build: String
    let os_version: String
    let device_model: String
    let locale: String
    let captured_at: String
    let include_logs: Bool
    let logs: String?
}

private nonisolated struct SubmitFeedbackParameters: Encodable, Sendable {
    let p_student_id: String
    let p_school_id: Int
    let p_category: String
    let p_message: String
    let p_platform: String
    let p_context: FeedbackContext
}

private nonisolated struct RegisterFeedbackAttachmentParameters: Encodable, Sendable {
    let p_feedback_id: String
    let p_kind: String
    let p_storage_path: String
    let p_mime_type: String
    let p_byte_size: Int
    let p_width: Int
    let p_height: Int
}
